### ISSUE TITLE
Allow installing from github

### ISSUE THREAD

**Issue by ndmitchell** [commercialhaskell/stack#1278]
I'd like to do:

    stack install https://github.com/ndmitchell/hoogle.git

And have that "just work". Bonus points if the version without `.git` also works. In particular, this would need to check out the repo, if a `stack.yaml` was missing it would need to `stack init`, and then `stack install`. See https://hackage.haskell.org/package/cabalg by @dmaliko for a version for Cabal.

**Comment by ndmitchell** [comment]
It would be nice if `stack install hoogle:git` worked too, since `stack` knows where the `hoogle` repo is.

**Comment by ndmitchell** [comment]
In fact, it would be super cool if `stack upgrade --git` was an alias for `stack install stack:git` - then everyone can reuse the upgrade stuff with very low effort.

**Comment by bitemyapp** [comment]
:+1: I like this :)

**Comment by mgsloan** [comment]
I agree that this would be cool!  It'd also be nice to be able to install from a URL / path to a tarball.

The `hoogle:git` syntax won't fly, because it's already used for the targets parsing - `hoogle:git` refers to building the `git` component (an exe, test, or bench).  If the parameter fails to be parsed as a package name / component spec, it's interpreted as a directory (and all packages that are contained within are included (recursively)).

So, targets syntax is already pretty involved.. But having a short thing for "Look up the source control metadata and use it" would be pretty darn sweet :) .  This would be like supporting `stack upgrade --git`, but for other projects than stack.  Could be a can of worms as well, though, if folks start asking for support for version control support beyond git.

**Comment by ndmitchell** [comment]
stack install hoogle --git is probably a better syntax, being consistent with upgrade. If you support --git and http tarballs then users of other VCS (those mythical people) can use a tarball of their repo.

**Comment by tchoutri** [comment]
Then perhaps should we think about a way to make `stack` more a package manager, with uninstalling and such?   
People (me included) are using it like that and if no (serious) alternative is provided to the community, more and more will complain that they cannot uninstall software with Stack.

**Comment by samvv** [comment]
I'm all for this 👍 I just found myself in a situation where I have developed a library that I want my installed tool ([Shake](https://shakebuild.com) to see. I currently see no way of doing that. Is there some kind of temporary workaround? Like copying/symlinking certain files?

**Comment by dbaynard** [comment]
Hi @samvv, more detail would be helpful. Is your library published somewhere?

In particular, the manual has a number of approaches that may be relevant — but from the information you've given so far it's hard to judge how to help.

**Comment by samvv** [comment]
@dbaynard Thanks for the response. [This](https://github.com/samvv/haskell-typed-command) is the library. It's isn't published anywhere yet; my account on Hackage has yet to get approved. I've managed to fix it by installing vanilla Cabal, doing a `stack sdist` and then a `cabal install <tarfile>`. A bit ugly, but it works.

**Comment by dbaynard** [comment]
That’s helpful. You may wish to use a custom snapshot which includes your github repo.

I can give more detail if the shakefile is public. Would you be willing to put it in a gist, or something?

**Comment by samvv** [comment]
The project is [over here](https:///github.com/samvv/AutoFFI), though it doesn't matter much, because I've switched back to CMake for this particular project instead of Shake. Thanks for wanting to help anyways, and I still think it would be a cool feature to have.
